invertedai/large/initialize.py

```diff
@@ -0,0 +1,296 @@
+    get_birdview: Optional[bool] = False,
+    get_infractions: Optional[bool] = False,
+    random_seed: Optional[int] = None,
+    api_model_version: Optional[str] = None,
+    display_progress_bar: Optional[bool] = True   
+) -> InitializeResponse:
+    """
+    A utility function to initialize an area larger than 100x100m. This function breaks up an 
+    area into a grid of 100x100m regions and runs initialize on them all. For any particular 
+    region of interest, existing agents in overlapping, neighbouring regions are passed as 
+    conditional agents to :func:`initialize`. Regions will be rejected and :func:`initialize` 
+    will not be called if it is not possible for agents to exist there (e.g. there are no 
+    drivable surfaces present) or if the agent density criterion is already satisfied. As well,
+    predefined agents may be passed via the regions and will be considered as conditional. 
+
+    Arguments
+    ----------
+    location:
+        Location name in IAI format.
+
+    regions:
+        List of regions that contains information about the location, shape, and number of agents within the region.
+    
+    traffic_light_state_history:
+        History of traffic light states - the list is over time, in chronological order, i.e.
+        the last element is the current state. If there are traffic lights in the map, 
+        not specifying traffic light state is equivalent to using iai generated light states.
+
+    get_birdview:
+        If True, a birdview image will be returned representing the current world. Note this will significantly
+        impact on the latency.
+
+    get_infractions:
+        If True, infraction metrics will be returned for each agent.
+    
+    random_seed:
+        Controls the stochastic aspects of initialization for reproducibility.
+
+    api_model_version:
+        Optionally specify the version of the model. If None is passed which is by default, the best model will be used.
+
+    display_progress_bar:
+        If True, a bar is displayed showing the progress of all relevant processes.
+    
+    See Also
+    --------
+    :func:`initialize`
+    """
+
+    agent_states_sampled = []
+    agent_attributes_sampled = []
+    agent_rs_sampled = []
+
+    def inside_fov(center: Point, agent_scope_fov: float, point: Point) -> bool:
+        return ((center.x - (agent_scope_fov / 2) < point.x < center.x + (agent_scope_fov / 2)) and
+                (center.y - (agent_scope_fov / 2) < point.y < center.y + (agent_scope_fov / 2)))
+    
+    if display_progress_bar:
+        iterable_regions = tenumerate(
+            regions, 
+            total=len(regions),
+            desc=f"Initializing regions"
+        )
+    else:
+        iterable_regions = regions
+
+    for i, region in iterable_regions:
+        region_center = region.center
+        region_size = region.size
+
+        existing_agent_states, existing_agent_attributes, _ = _get_all_existing_agents_from_regions(regions)
```

Comment:
I think you can pull this out of the loop?

Reply:
The regions are populated with agents one at a time. Therefore, each loop all agents currently in the simulation (initialized, preddefined, whatever) need to be accessed so that ones that are relevant to the current region of interest can be included as conditional. An alternative would be to have 3 running lists for states, attributes, and recurrent states but I went with this method for the sake of saving memory space and code cleanliness at the cost of a small increase in computation time due to looping through every region.